### ISSUE TITLE
Seeking reasons for discrepancies between experimental results and official records

### ISSUE THREAD

**Issue by LL-Zs** [ubicomplab/rPPG-Toolbox#294]
Hello, we used the officially released source code and yaml file, but the result data from the training is the same, the results from changing different learning rates are still the same, and the results we finally get are somewhat different from the officially released results. What is the reason? The following are the results we obtained using official source code experiments: UBFC-PURE. yaml, 4.96 (official 3.69), 4.79 (official 3.38). UBFC-MMPD, 1.54 (official 1.46), 2.17 (official 2.04). PURE-MMPD 2.07 (official 1.78), 2.92 (official 2.47)

**Comment by yahskapar** [comment]
Hi @LL-Zs,

Can you elaborate on what exactly you did, and include additional details such as what GPUs you used and the exact `.yaml` config files you used unless you are certain you didn't change the default settings aside from the learning rate parameter you mentioned you varied? Maybe do this as a numbered or bulleted list to make it a bit easier to understand the whole picture.

It is not trivial to reproduce training results sometimes, especially across different versions of key libraries (e.g., torch), different GPU architectures, and considering there may have been some changes to the repo with various parts of the pipeline (e.g., preprocessing involving face detection) that may differ significantly than the version of the toolbox used to capture the results in the late-2023, NeurIPS version of the rPPG-Toolbox paper. Note that pre-trained models used for the paper can be found [here](https://github.com/ubicomplab/rPPG-Toolbox/tree/main/final_model_release). Hopefully you are not noticing any significant discrepancy with those pre-trained models (if you do, please let us know).

**Comment by LL-Zs** [comment]
Thank you for your reply. We downloaded three datasets, PURE, UBFC, and MMPD, and directly reproduced them using the official code of Toolbox (both the dataset and code are the latest downloaded, and the code is the end of 2023 version). The reproduction process includes "train and test" and using the PTH file "only test" from the final mode release in the official code. Here is our configuration situation:
GPU：NVIDIA GeForce RTX 3090
CPU: Intel Xeon (R) Silver 4210R processor
Operating System: Ubuntu 20.04.6 LTS
Python version: 3.8.19
Torch version: 2.3.1
CUDA version: 11.8
YAML file and corresponding results:
PURE-PURE-UBFC rppg TSCAN - "train and test" -1.44, 1.63- final pth "only test" -1.29, 1.50.
UBFC-rppg-UBFC-rppg-PURE-TSCAN——"train and test"-4.97,4.80——final pth "only test" -3.65,3.32.
In the MMPD dataset test, the results obtained were also slightly higher than those of the official strategy.
During the replication process, the code was downloaded directly without changing any other configurations.

**Comment by yahskapar** [comment]
Hi @LL-Zs,

For the `only_test` results that use the pre-trained models, these results seem almost identical. Your own training results don't really seem that crazy either, and could be totally reasonable given the hardware differences. What exactly are you worried about here? 

I think for the point of publication, if you find that re-training on different hardware to not match existing benchmark results closely enough, just use your results as the baseline for comparisons. As long as you note the hardware differences (especially the GPU used) and include more robust measurement details such as the standard error, I don't see how this could be a problem.

**Comment by yahskapar** [comment]
Closing this due to a lack of further discussion, but please feel free to re-open or make a new issue if needed @LL-Zs.

**Comment by Ishtiaque-khan97** [comment]
Hi, @yahskapar 
just wanted to ask pretty much the same question. Is it expected for the training results to vary from the paper results?
For example, I used this config file for training on UBFC-Rppg and testing on UBFC-PHYS : 
"UBFC-rPPG_UBFC-rPPG_UBFC-PHYS_DEEPPHYS_BASIC.yaml"

I got these results:
```
FFT MAE (FFT Label): 7.2564153387995045 +/- 0.9264204903763117
FFT RMSE (FFT Label): 11.804207342852903 +/- 30.101568352398264
FFT MAPE (FFT Label): 8.868698705184176 +/- 1.0928499460291792
FFT Pearson (FFT Label): 0.6053787004894869 +/- 0.07999473650574035
FFT SNR (FFT Label): -3.088645639566399 +/- 0.8261550800622746 (dB)

```
These are a bit different from the results in the paper. For example, the MAE was reported as  6.62±0.84. Is such discrepancy expected? I want to make sure that I am using the toolbox and the config files properly.

These are the subjects/videos in the UBFC-rppg folder:

`subject1   subject12  subject15  subject18  subject23  subject26  subject30  subject33  subject36  subject39  subject41  subject44  subject47  subject5 subject10  subject13  subject16  subject20  subject24  subject27  subject31  subject34  subject37  subject4   subject42  subject45  subject48  subject8 subject11  subject14  subject17  subject22  subject25  subject3   subject32  subject35  subject38  subject40  subject43  subject46  subject49  subject9`

I am attaching the config file here:

[UBFC-rPPG_UBFC-rPPG_UBFC-PHYS_DEEPPHYS_BASIC.txt](https://github.com/user-attachments/files/18594536/UBFC-rPPG_UBFC-rPPG_UBFC-PHYS_DEEPPHYS_BASIC.txt)

**Comment by yahskapar** [comment]
Hi @Ishtiaque-khan97,

See my previous replies in this thread - as I mentioned, there are numerous factors, including hardware differences, that could contribute to some (usually minor) discrepancies in results. In your case, I don't think the discrepancy is too large and your config file itself looks fine. 

One other thing to check is the performance of the pre-trained model on your machine - try that with [this config](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/configs/infer_configs/UBFC-rPPG_UBFC-PHYS_DEEPPHYS_BASIC.yaml). You may get the same result as the paper (e.g., MAE of 6.62±0.84), or perhaps something slightly different and maybe even closer to the results you got from your training run. If some concerns of yours persist on this topic, by the way, please try to make a new issue.

All the best,

Akshay
